content/2018/12/10/channels-and-drf.rst

```diff
@@ -0,0 +1,292 @@
+OddBird, so I thought I'd write up an in-progress report. I'm not at all
+convinced that this is the best way, but it's what seems to be working
+so far.
+
+The basics
+----------
+
+First, DRF. If you're here, you're probably already familiar with it.
+We've made a pretty traditional RESTful API with it, keeping the
+endpoints flat, with minimal server-side logic mostly encapsulated in
+the serializers.
+
+The endpoints look a bit like this::
+
+   GET /api/foo
+   POST /api/foo
+   GET /api/foo/:id
+   PUT /api/foo/:id
+   DELETE /api/foo/:id
+
+And the code like this::
+
+   foo/
+     __init__.py
+     models.py
+     serializers.py
+     urls.py
+     views.py
+     tests/
+       __init__.py
+       models.py
+       serializers.py
+       views.py
+
+As a side note, for something that I'm not distributing as a library, I
+like to keep the tests parallel to the code and within the same tree; if
+I'm writing a library, I root the tests in a different tree, but still
+with parallel structure to the code. This makes it easier to exclude
+them on an install.
+
+Inside those files, we mostly have simple declarative use of DRF. Follow
+their `tutorial`_ if you want to get that set up. We run our tests with
+`pytest`_ and a requirement of 100% test coverage, as we do with all our
+projects.
+
+Channels is perhaps a little less likely to be familiar; I haven't seen
+it get quite as much uptake, being a more specialized tool. That said,
+it's pretty simple to set up on top of an existing Django application,
+so I'll walk you through it in brief.
+
+First, you need to move from using WSGI to ASGI, which is "basically the
+same, but async". This means changing your server process from gunicorn
+to something like Daphne, changing your ``project.wsgi`` file to
+``project.asgi`` (as described in the `Channels docs`_), add a
+``routing`` module and a ``consumers`` module, and adjust your settings
+appropriately.
+
+At this stage, you won't yet have anything in ``consumers``, and you
+won't have much in routing. It can look like this::
+
+
+   from channels.routing import ProtocolTypeRouter
+
+   application = ProtocolTypeRouter({
+       # (http->django views is added by default)
+   })
+
+Yep, that's basically an empty ``ProtocolTypeRouter``. We're just making
+sure we don't break anything with the transition to ASGI, first, and
+``ProtocolTypeRouter`` does as the comment says, and wires HTTP to
+Django by default.
+
+Once that's all done and you've confirmed that everything's still
+working, you can start to add in the wiring for WebSockets.
+
+WebSockets, Consumers, and Groups
+---------------------------------
+
+Let's talk a bit about architecture before we talk about how to
+implement it.
+
+As we're using it, Channels primarily drives the pushing of
+notifications about the state of selected objects to the client, over
+WebSockets. We've opted to simplify the client's job by having one
```

Comment:
I find this first sentence difficult to parse. What exactly is channels driving? Is it the full clause "the pushing of
notifications about the state of selected objects to the client"?